src/CameraOnvif.cpp

```diff
@@ -0,0 +1,369 @@
+        reportError();
+    }
+    response.color_saturation = *get_imaging_resp.ImagingSettings->ColorSaturation;
+    response.brightness = *get_imaging_resp.ImagingSettings->Brightness;
+    response.contrast = *get_imaging_resp.ImagingSettings->Contrast;
+
+    return response;
+}
+
+void CameraOnvif::setImageParam(const ImageParam& new_param){
+    float new_brightness = (m_brightness[1] - m_brightness[0]) *
+                            new_param.brightness + m_brightness[0];
+    float new_color_saturation = (m_color_saturation[1] - m_color_saturation[0]) *
+                                  new_param.color_saturation + m_color_saturation[0];
+    float new_contrast = (m_contrast[1] - m_contrast[0]) *
+                          new_param.contrast + m_contrast[0];
+
+    _timg__SetImagingSettings set_imaging;
+    set_imaging.VideoSourceToken = m_video_token;
+    set_imaging.ImagingSettings = new tt__ImagingSettings20();
+    set_imaging.ImagingSettings->Brightness = &new_brightness;
+    set_imaging.ImagingSettings->ColorSaturation = &new_color_saturation;
+    set_imaging.ImagingSettings->Contrast = &new_contrast;
+
+    _timg__SetImagingSettingsResponse set_imaging_resp;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&set_imaging, set_imaging_resp)) {
+        reportError();
+    }
+}
+
+void CameraOnvif::setResolution(int width, int height){
+    bool valid = false;
+    for (auto & e : m_private->video_options.Options->H264->ResolutionsAvailable) {
+        if (e->Width == width && e->Height == height) {
+            valid = true;
+        }
+    }
+    if (!valid) {
+        throw SOAPException("You are trying to set a video resolution that is not available!");
+    }
+
+    _trt__GetVideoEncoderConfigurationResponse conf_get_response;
+    getVideoEncoderConfiguration(conf_get_response);
+
+    _trt__SetVideoEncoderConfiguration conf_set;
+    _trt__SetVideoEncoderConfigurationResponse conf_set_response;
+    conf_set.Configuration = conf_get_response.Configuration;
+
+    conf_set.Configuration->Resolution->Width = width;
+    conf_set.Configuration->Resolution->Height = height;
+    setCredentials();
+    if (m_private->proxy_media->SetVideoEncoderConfiguration(&conf_set, conf_set_response)){
+        reportError();
+    }
+}
+
+void CameraOnvif::getVideoEncoderConfiguration(
+    _trt__GetVideoEncoderConfigurationResponse& resp
+) {
+    _trt__GetVideoEncoderConfiguration GetVideoEncoderConfiguration;
+    GetVideoEncoderConfiguration.ConfigurationToken = m_video_conf_token;
+
+    setCredentials();
+    bool error = m_private->proxy_media->GetVideoEncoderConfiguration(
+        &GetVideoEncoderConfiguration, resp);
+    if (error) {
+        reportError();
+    }
+}
+
+camera_onvif::Resolution CameraOnvif::getResolution(){
+    _trt__GetVideoEncoderConfigurationResponse video_conf;
+    getVideoEncoderConfiguration(video_conf);
+    Resolution response = Resolution();
```

Comment:
No need to explicitly initialize with default object. `Resolution response;` already calls the default constructor.